File: python_client/src/ezauth_client/tables.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ._client import _encode
from .rows import Rows

if TYPE_CHECKING:
    from ._client import BaseClient


class Columns:
    def __init__(self, client: BaseClient):
        self._client = client

    def add(
        self,
        table_id: str,
        name: str,
        type: str,
        *,
        required: bool = False,
        default_value: str | None = None,
        position: int | None = None,
    ) -> dict:
        return self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/columns",
            method="POST",
            body={
                "name": name,
                "type": type,
                "required": required,
                "default_value": default_value,
                "position": position,
            },
        )

    def update(
        self,
        table_id: str,
        column_id: str,
        *,
        name: str | None = None,
        required: bool | None = None,
        default_value: str | None = ...,
        position: int | None = None,
    ) -> dict:
        body: dict = {}
        if name is not None:
            body["name"] = name
        if required is not None:
            body["required"] = required
        if default_value is not ...:
            body["default_value"] = default_value
        if position is not None:
            body["position"] = position
        return self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/columns/{_encode(column_id)}",
            method="PATCH",
            body=body,
        )

    def delete(self, table_id: str, column_id: str) -> None:
        self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/columns/{_encode(column_id)}",
            method="DELETE",
        )


class Tables:
    def __init__(self, client: BaseClient):
        self._client = client
        self.columns = Columns(client)
        self.rows = Rows(client)

    def create(self, name: str, *, columns: list[dict] | None = None) -> dict:
        body: dict = {"name": name}
        if columns is not None:
            body["columns"] = columns
        return self._client._fetch("/v1/tables", method="POST", body=body)

    def list(self) -> dict:
        return self._client._fetch("/v1/tables")

    def get(self, table_id: str) -> dict:
        return self._client._fetch(f"/v1/tables/{_encode(table_id)}")

    def delete(self, table_id: str) -> None:
        self._client._fetch(f"/v1/tables/{_encode(table_id)}", method="DELETE")

File: swift_client/Sources/EZAuth/Tables.swift
import Foundation

/// Table and column CRUD.
public struct Tables: Sendable {
    let client: BaseClient

    /// Column management sub-namespace.
    public var columns: Columns { Columns(client: client) }

    /// Row management sub-namespace.
    public var rows: Rows { Rows(client: client) }

    // MARK: - Create table

    public func create(name: String, columns: [ColumnDefinition]? = nil) async throws -> TableDetailResponse {
        try await client.fetch(
            TableDetailResponse.self,
            path: "/v1/tables",
            method: "POST",
            body: CreateTableRequest(name: name, columns: columns),
            auth: .auto
        )
    }

    // MARK: - List tables

    public func list() async throws -> TableListResponse {
        try await client.fetch(
            TableListResponse.self,
            path: "/v1/tables",
            auth: .auto
        )
    }

    // MARK: - Get table

    public func get(_ tableId: String) async throws -> TableDetailResponse {
        try await client.fetch(
            TableDetailResponse.self,
            path: "/v1/tables/\(tableId.urlPathEncoded)",
            auth: .auto
        )
    }

    // MARK: - Delete table

    public func delete(_ tableId: String) async throws {
        try await client.fetch(
            path: "/v1/tables/\(tableId.urlPathEncoded)",
            method: "DELETE",
            auth: .auto
        )
    }
}

// MARK: - Columns sub-namespace

public struct Columns: Sendable {
    let client: BaseClient

    public func add(
        tableId: String,
        name: String,
        type: String,
        required: Bool = false,
        defaultValue: JSONValue? = nil,
        position: Int? = nil
    ) async throws -> ColumnResponse {
        try await client.fetch(
            ColumnResponse.self,
            path: "/v1/tables/\(tableId.urlPathEncoded)/columns",
            method: "POST",
            body: CreateColumnRequest(
                name: name,
                type: type,
                required: required,
                default_value: defaultValue,
                position: position
            ),
            auth: .auto
        )
    }

    public func update(
        tableId: String,
        columnId: String,
        name: String? = nil,
        required: Bool? = nil,
        defaultValue: JSONValue? = nil,
        position: Int? = nil
    ) async throws -> ColumnResponse {
        try await client.fetch(
            ColumnResponse.self,
            path: "/v1/tables/\(tableId.urlPathEncoded)/columns/\(columnId.urlPathEncoded)",
            method: "PATCH",
            body: UpdateColumnRequest(
                name: name,
                required: required,
                default_value: defaultValue,
                position: position
            ),
            auth: .auto
        )
    }

    public func delete(tableId: String, columnId: String) async throws {
        try await client.fetch(
            path: "/v1/tables/\(tableId.urlPathEncoded)/columns/\(columnId.urlPathEncoded)",
            method: "DELETE",
            auth: .auto
        )
    }
}

// MARK: - Request types

struct CreateTableRequest: Encodable {
    let name: String
    let columns: [ColumnDefinition]?
}

/// Column definition used when creating a table.
public struct ColumnDefinition: Encodable, Sendable {
    public let name: String
    public let type: String
    public let required: Bool
    public let default_value: JSONValue?
    public let position: Int?

    public init(name: String, type: String, required: Bool = false, defaultValue: JSONValue? = nil, position: Int? = nil) {
        self.name = name
        self.type = type
        self.required = required
        self.default_value = defaultValue
        self.position = position
    }
}

struct CreateColumnRequest: Encodable {
    let name: String
    let type: String
    let required: Bool
    let default_value: JSONValue?
    let position: Int?
}

struct UpdateColumnRequest: Encodable {
    let name: String?
    let required: Bool?
    let default_value: JSONValue?
    let position: Int?
}

// MARK: - Response types

public struct TableResponse: Decodable, Sendable {
    public let id: String
    public let name: String
    public let created_at: String
    public let updated_at: String
}

public struct TableListResponse: Decodable, Sendable {
    public let tables: [TableResponse]
    public let total: Int
}

public struct TableDetailResponse: Decodable, Sendable {
    public let id: String
    public let name: String
    public let columns: [ColumnResponse]
    public let created_at: String
    public let updated_at: String
}

public struct ColumnResponse: Decodable, Sendable {
    public let id: String
    public let name: String
    public let type: String
    public let required: Bool
    public let default_value: JSONValue?
    public let position: Int
    public let created_at: String
    public let updated_at: String
}
